xarray/core/dataset.py

```diff
@@ -7007,5 +7008,235 @@ def argmax(self, dim=None, axis=None, **kwargs):
                 "Dataset.argmin() with a sequence or ... for dim"
             )
 
+    def curvefit(
+        self,
+        coords: Union["DataArray", Iterable["DataArray"]],
+        func: Callable[..., Any],
+        reduce_dim: Union[Hashable, Iterable[Hashable]] = None,
+        skipna: bool = True,
+        cov: bool = False,
+        p0: Dict[str, Any] = None,
+        bounds: Dict[str, Any] = None,
+        param_names: Sequence[str] = None,
+        kwargs: Dict[str, Any] = None,
```

Comment:
```suggestion
        reduce_dim: Union[Hashable, Iterable[Hashable]] = [],
        skipna: bool = True,
        cov: bool = False,
        p0: Dict[str, Any] = {},
        bounds: Dict[str, Any] = {},
        param_names: Sequence[str] = None,
        kwargs: Dict[str, Any] = {},
```
Can't we have better default values instead to avoid all those None checks?

Reply:
This is not a good idea: https://docs.python-guide.org/writing/gotchas/#what-you-wrote

Reply:
I defer to the experts. This is my first PR here, so was trying to copy the typing behavior from other functions.

Reply:
Also welcome any other feedback on proper type hinting. For example, I think we should actually allow:
```python
coords: Union[Union[str, "DataArray"], Iterable[Union[str, "DataArray"]]]
```
But this raises errors in mypy:
```
xarray/core/dataset.py:7108: error: Item "str" of "Union[str, DataArray]" has no attribute "equals"
xarray/core/dataset.py:7121: error: Item "str" of "Union[Any, str, DataArray]" has no attribute "broadcast_like"
```
Even though any strings passed to `coords` get converted to `self[coord]`.

Reply:
@dcherian That's interesting. It feels like I've used lists and dicts as defaults so many times and yet I don't remember this quirk. Maybe I just never run those functions more than once...

Reply:
mypy does not like the redefining `List[str] -> List[DataArray]` on line 7104. You'll need to rename the variable, e.g. `coords_` or we'd need to [add a mypy flag](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-allow-redefinition).

Reply:
Thought I had tried that approach (renaming `coords` in 7104), but that does indeed make mypy happy. Thanks.